Experimental: Add Clifford algebras and Clifford orders

SirToby25 opened the pull request:
This is the first approach of implementing Clifford algebras and Clifford orders for quadratic spaces and lattices. There will be some future development, but it already allows for the standard computations one might want to perform within these structures.

Resolves #2312

SirToby25 committed:
Setting up Clifford algebras

Provides Basic functionality for Clifford algebras over free quadratic
lattices. Still a lot of work to do: Implement Clifford orders, handle
pseudo-bases in the non-free case, etc.

SirToby25 committed:
Updated Typenames

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Backup commit

SirToby25 committed:
Updated type variables in the data structure

SirToby25 committed:
Center, centroid and quadratic discriminant added

SirToby25 committed:
Minor changes I do not recall

SirToby25 committed:
Updated testfile to new data structure

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Data structure for Clifford orders and testfile

This commit also contains minor changes to the CliffordAlgebras-files,
mainly due to consistency with the CliffordOrders-files

SirToby25 committed:
I messed up

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Finished data structure for Clifford orders

Clifford orders are now fully functional. Some methods for Clifford
algebras were renamed for consistency. Consequently the tests for
Clifford algebras were modified as well.

SirToby25 committed:
Updated the various divexact methods

divexact methods were implemented for Clifford orders. Also updated the
existing one for Clifford algebras for consistency.

SirToby25 committed:
Added tests for divexact methods

SirToby25 committed:
Added test for base_ring_type

SirToby25 committed:
Changed constructor for corner cases

SirToby25 committed:
Wrote first tests for Clifford orders

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Used JuliaFormatter

SirToby25 committed:
Minor bugfixes

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Bug fix in _set_centroid_and_disq!

SirToby25 committed:
Added new example

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merging

Merge branch 'tb/CliffordAlgebras' of github:SirToby25/Oscar.jl into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
getindex and tests for Clifford algebras

This commit adds a getindex method for the elements of a Clifford
algebra so that its entries can be accessed directly

SirToby25 committed:
Getindex for Clifford orders and further examples

This commit adds a getindex method for Clifford orders and some tests.
Also, further test examples are part of this commit.

SirToby25 committed:
Fixed doc for 'quadratic_discriminant'

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Removed unnecessary exports

SirToby25 committed:
Removed unecessary specifications of functions

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+export CliffordAlgebra,
```
IMO this could be put into one single experimental project. This way, we would not need to care about the correct inclusion order of the two experimental folders (which is an issue atm)

lgoettgens commented:
IMO this could be put into one single experimental project. This way, we would not need to care about the correct inclusion order of the two experimental folders (which is an issue atm).

Furthermore, I think it would be great if @thofma could comment if this should go into Oscar experimental or Hecke. I noticed that this is dependent on some internals of `Hecke.AbstractAssociativeAlgebra`, and thus having this here could make further development in Hecke more challenging.

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+
+parent_type(::Type{CliffordAlgebraElem{T,S,P}}) where {T,S,P} = P
+
+base_ring_type(C::CliffordAlgebra{T,S}) where {T,S} = typeof(base_ring(C))
```
`base_ring_type` should take `Type{CliffordAlgebra}` as its argument and should be computable at compile time. I think this could just be `parent_type(T)` here, but not sure

SirToby25 replied:
Fixed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/73f677cfe5c0362d5f9763d3d3b24a6e832e63a2)

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+
+Return the base ring of the Clifford algebra $C$.
+"""
+base_ring(C::CliffordAlgebra) = C.base_ring
```
```suggestion
base_ring(C::CliffordAlgebra) = C.base_ring::TypeAnnotation
```
this needs a type annotation, to not infer `Ring` here. You should be able to just put your updated implementation of `base_ring_type` there

fingolfin replied:
i.e.
```suggestion
base_ring(C::CliffordAlgebra) = C.base_ring::base_ring_type(typeof(C))
```

SirToby25 replied:
Fixed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/73f677cfe5c0362d5f9763d3d3b24a6e832e63a2)

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+
+Return the base ring of the Clifford algebra $C$.
+"""
+base_ring(C::CliffordAlgebra) = C.base_ring
```
```suggestion
base_ring(C::CliffordAlgebra) = C.base_ring::TypeAnnotation
```
this needs a type annotation, to not infer `Ring` here. You should be able to just put your updated implementation of `base_ring_type` there

thofma commented:
I will have a look over the weekend.

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+Return the coefficient vector of $x$ wrt the
+canonical basis of its parent Clifford algebra.  
+"""
+coeff(x::CliffordAlgebraElem) = x.coeffs
```
Iirc, `coeff` usually takes an additional index and returns the corresponding coefficient. The function returning all coefficients is then called `coefficients`

SirToby25 replied:
This is taken care of in [commit f9eb179](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/f9eb1797ffbcefa74b3d134480af6fc6a70eafa2)

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,579 @@
+#
+################################################################################
+
+Base.:+(x::CliffordAlgebraElem) = x
```
```suggestion
```
unary + is handled generically in AbstractAlgebra

lgoettgens reviewed (commented):
some misc comments about the first part

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Update experimental/CliffordAlgebras/src/CliffordAlgebras.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

SirToby25 committed:
Renamed 'coeff' to coefficients. Added 'coeff'

The previous 'coeff' method is now called 'coefficients'. Also there is
now a method 'coeff' available that takes an Int as second argument and
returns the i-th entry of coefficients.

thofma commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+### Elements ###
+(C::CliffordAlgebra)() = CliffordAlgebraElem(C)
+
+function (C::CliffordAlgebra)(a::R) where {R<:RingElem}
```
```suggestion
function (C::CliffordAlgebra)(a::R) where {R<:RingElement}
```
See
```
julia> RingElement
RingElement (alias for Union{RingElem, AbstractFloat, Integer, Rational})
```
Maybe this includes all the things you want and you can remove the `Number` version below?

SirToby25 replied:
I have tried this in the past but it doesn't work. In several tests I get:

MethodError: (::CliffordAlgebra{AbsSimpleNumFieldElem, AbstractAlgebra.Generic.MatSpaceElem{AbsSimpleNumFieldElem}})(::AbsSimpleNumFieldElem) is ambiguous

MethodError: (::CliffordAlgebra{QQFieldElem, QQMatrix})(::QQFieldElem) is ambiguous

Edit: Your suggestions work for the "divexact" methods. Thanks

thofma commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+@doc raw"""
+    gens(C::CliffordAlgebra) -> Vector{CliffordAlgebraElem}
+
+Return the vector of canonical multiplicative generators of the Clifford algebra $C$.
```
What are multiplicative generators? Generators of the unit group of C?

For `AbstractAssociativeAlgebra` this returns algebra generators (over the base field). If this here is doing something else, the name needs to be different.

SirToby25 replied:
Description clarified in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+is odd. Otherwise it is one-dimensional and returned as a tuple of length one
+containing the multiplicative identity of $C$.
+"""
+function center(C::CliffordAlgebra)
```
For algebras this should return a subalgebra together with the embedding. If you want to return a basis, maybe rename it to `center_basis` or `basis_of_center`?

SirToby25 replied:
Done in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+
+  function ZZCliffordOrder(ls::ZZLat)
+    if !is_zero(rank(ls))
+      @req is_integral(1//2 * norm(ls)) "The given lattice is not even!"
```
Isn't there an `is_even(ls)` check?

SirToby25 replied:
Yes there is. I will change it, thanks. Then this method is missing only for QuadLat

SirToby25 replied:
Fixed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/73f677cfe5c0362d5f9763d3d3b24a6e832e63a2)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+#
+#############################################################
+
+mutable struct CliffordOrder{T, C} <: Hecke.AbstractAssociativeAlgebra{T}
```
I think all `AbstractAssociativeAlgebra` need to be free over the base ring, to make use of some common functionality. How about making it a subtype of `Hecke.AlgAssRelOrd`?

SirToby25 replied:
That's a good idea

SirToby25 replied:
I would like to but `Hecke.AlgAssRelOrd` is not an abstract type. Is this intended?

```julia
julia> isabstracttype(Hecke.AlgAssRelOrd)
false

thofma replied:
Sorry, I was hallucinating. Indeed there is no abstract type. I just made them 
```
julia> supertype(Hecke.AlgAssRelOrd)
NCRing
```
since they are not necessarily free.

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+  end
+end
+
+mutable struct ZZCliffordOrder <: Hecke.AbstractAssociativeAlgebra{ZZRingElem}
```
Similar suggestion as above with `Hecke.AlgAssAbsOrd`.

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+
+Return the Clifford algebra of the ambient space of the underlying lattice of $C$.
+"""
+ambient_algebra(C::ZZCliffordOrder) = C.ambient_algebra
```
I think for the other orders in Oscar, this is just called `algebra`.

SirToby25 replied:
Changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+  ba = base_ring(ambient_algebra(C))
+  res = zero_matrix(ba, 1, rank(C))
+  res[1, i] = ba(1)
+  return pseudo_matrix(res, [coefficient_ideals(C)[i]])
```
Hm, shouldn't a pseudo-basis consist of pairs of elements of the algebra and coefficient ideals? At least this is how this terminology is used everywhere else in Oscar.

thofma replied:
For example, for lattices in quadratic spaces over number fields, it is a tuple `(Vector, Ideal)`. For orders in relative number fields it is `(NumFieldElem, Ideal)`.

SirToby25 replied:
Changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+
+Return the canonical pseudo-basis of the Clifford order $C$.
+"""
+pseudo_basis(C::CliffordOrder) = pseudo_matrix(identity_matrix(base_ring(ambient_algebra(C)), rank(C)), coefficient_ideals(C))
```
This would be a pseudo-basis matrix.

SirToby25 replied:
Output type changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+
+Return the canonical multiplicative pseudo-generating set of the Clifford order $C$.
+"""
+function gens(C::CliffordOrder) 
```
If it is not a generating set, it should not be called `gens`.

SirToby25 replied:
Renamed to pseudo_gens and output type changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma commented on experimental/CliffordOrders/src/CliffordOrders.jl:
```diff
@@ -0,0 +1,1056 @@
+Return the center of $C$. It equals `centroid(C)`, if and only if `rank(lattice(C))`
+is odd. Otherwise it free of rank one, generated by the multiplicative identity of $C$.
+"""
+function center(C::CliffordOrder)
```
As above for the algebra.

SirToby25 replied:
Renamed to pseudo_basis_of_center in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/d4b78c05016fbec7d6ab36f6b18b4b8bec2d084d)

thofma reviewed (commented):
This all looks very nice. I don't mind if this lands in Oscar or Hecke. If the aim is to integrate this more with representation theory, it makes sense to have it here.

I have left a few comments, in particular concerning the names of some methods. In some cases they are not in agreement with the convention that we have so far. I am also happy to discuss this on slack if you have any questions or concerns.

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+    else
+      x.parent.base_ring()
+    end, 1:(x.parent.dim)
+  )
```
```suggestion
  R = base_ring(parent(x))
  d = dim(parent(x))
  x.even_coeffs = [ iseven(count_ones(y - 1)) ? x.coeffs[y] : R() for y in 1:d]
```

fingolfin replied:
And of course similar in the other `_set_even_odd_coefficients` method.

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+################################################################################
+
+function Base.:+(x::CliffordAlgebraElem{T}, y::CliffordAlgebraElem{T}) where {T<:FieldElem}
+  @req parent(x) === parent(y) "The inputs must lie in the same Clifford algebra"
```
```suggestion
  check_parent(x, y)
```
Same everywhere else

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+  map(
+    j ->
+      x[j] .*
+      _mul_baseelt_with_gen(BitVector(digits(j - 1; base=2, pad=ncols(gram))), i, gram),
```
Converting `Int` values into `BitVector` seems very inefficient to me. Can't we do we without it?

SirToby25 replied:
Thanks, of course we can. Changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+#Converts a bitvector to an integer. The first entry corresponds to 2^0, the second one to 2^1 and so on.
+function _bitvec_to_int(x::BitVector)
+  pow_of_two = 1
+  res = 0
```
So you convert to `Int`... so at most 64 bits are allowed in your bit vector `x`.

So why do you use a `BitVector` at all?

SirToby25 replied:
`BitVector` is not used anymore as of this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+function _mul_baseelt_with_gen(x::BitVector, i::Int, gram::MatElem)
+  R = base_ring(gram)
+  j = _get_last(x)
+  res = fill(R(), 2^length(x))
```
In the call to this function `x` is padded to `ncols(gram)`, so presumably this would just be the same as this (??)
```suggestion
  res = fill(R(), 2^ncols(gram))
```

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+function _shift_entries!(X::Vector, s::Int)
+  X[(s + 1):end] = X[1:(end - s)]
+  X[1:s] = fill(parent(X[1])(), s)
+  return X
```
Isn't this the same as this?
```suggestion
  R = parent(X[1])
  return prepend!(X, fill(R(), s))
```
(In which case I'd suggest to remove this function and use `prepend!` directly).

SirToby25 replied:
Not quite, because the length of the vector `X` should not change.

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+#'i'-th generator of the Clifford algebra containing said basis vector.
+function _mul_baseelt_with_gen(x::BitVector, i::Int, gram::MatElem)
+  R = base_ring(gram)
+  j = _get_last(x)
```
If `x` was directly an `Int` then I think this line could just become
```suggestion
  j = 8*sizeof(Int) - leading_zeros(x)
```

SirToby25 replied:
`_get_last` became obsolete with this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b) and is not used anymore.

Edit: I adopted your brilliant computation of `j`, which does not use `log2`. Thank you!

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+  j = _get_last(x)
+  res = fill(R(), 2^length(x))
+  if j < i
+    res[_get_basis_index(x) + 2^(i - 1)] = R(1)
```
If `x` was an `Int` then this could become
```suggestion
    res[x + 2^(i - 1)] = R(1)
```

SirToby25 replied:
Done in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+    return res
+  end
+  if j == i
+    res[_get_basis_index(x) - 2^(i - 1)] = R(divexact(gram[i, i], 2))
```
If `x` was an `Int` then this could become
```suggestion
    res[x - 2^(i - 1)] = R(divexact(gram[i, i], 2))
```

SirToby25 replied:
Done in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).

fingolfin commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+  end
+  xj_to_zero = copy(x)
+  xj_to_zero[j] = 0
+  res[_get_basis_index(xj_to_zero)] = gram[i, j]
```
```suggestion
  xj_to_zero = xor(x, 2^(j-1))
  res[xj_to_zero] = gram[i, j]
```
or equivalently
```suggestion
  xj_to_zero = x - 2^(j-1)
  res[xj_to_zero] = gram[i, j]
```

SirToby25 replied:
Done in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).

fingolfin reviewed (commented):
Thank you for your contribution, which is much appreciated! I left a few technical comments. These are not meant to take away or block your overall project, but perhaps some of them are helpful.

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Several small fixes

-fixed the base_ring_type methods
-added a type assertion fo the base_ring methods
-updated divexact methods to work with RingElements

lgoettgens commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,584 @@
+
+### Elements ###
+mutable struct CliffordAlgebraElem{T,S,P} <: Hecke.AbstractAssociativeAlgebraElem{T}
+  parent::P
```
I think you can get rid of the third type param here and instead use
```suggestion
  parent::CliffordAlgebra{T, S}
```
This would make the code a bit simpler and thus easier to understand and maintain

SirToby25 replied:
Done in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/335fc8a255bb8159870818f9ac437c04dd832981).

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Renamed various methods and changed output types

Many methods now have a name that clarifies their output:
-'centroid' is now called 'basis_of_centroid' (or
'pseudo_basis_of_centroid')
-'center' is now called 'basis_of_center' (or 'pseudo_basis_of_center')
- For CliffordOrders, 'gen' and 'gens' is replaced by 'pseudo_gen' and
  'pseudo_gens'
-'ambient_algebra' is now called 'algebra'

All methods containing 'pseudo_basis' in their name (e.g.
'pseudo_basis', 'pseudo_basis_of_center', etc.) now return a vector of
tuples instead of a pseudo-matrix. This is consistent with similar
methods in Oscar.

The methods 'gen' and 'gens' have updated documentation that clarifies
what is meant by a generator.

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

thofma commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,589 @@
+function gen(C::CliffordAlgebra, i::Int)
+  res = C()
+  if i <= 0
+    res[0] #Throws a BoundsError instead of a DomainError for consistency
```
Out of curiosity: Consistency with respect to what?

SirToby25 replied:
Consistency in the sense that I think that

```julia

julia> gen(C, 0)
julia> gen(C, 3)
```
should both throw a `BoundsError`, provided the algebra `C` has only two generators

SirToby25 committed:
Using iseven and countones I/II

Co-authored-by: Max Horn <max@quendi.de>

thofma commented on experimental/CliffordAlgebras/src/CliffordAlgebras.jl:
```diff
@@ -0,0 +1,589 @@
+  return x
+end
+
+function __can_convert_coefficients(coeff::Vector{R}, K::Field) where {R}
```
(We need a function to do this. I have written such a function a dozen of times now.)

SirToby25 committed:
Using iseven and countones II/II

SirToby25 committed:
Using check_parent I/II

Co-authored-by: Max Horn <max@quendi.de>

thofma commented:
All my questions and comments have been addressed, thanks @SirToby25! I think from the interface/names viewpoint this PR is good to go.

lgoettgens commented:
https://github.com/oscar-system/Oscar.jl/pull/4370#discussion_r1863701951 is not addressed yet. Either the two files need to be but in a single experimental folder, or the dependency has to be put into `Experimental.jl`

thofma commented:
(I agree that one folder would be suitable.)

SirToby25 committed:
Using check_parents II/II

Replaced @req's by check_parents whenever suitable. Fixed the '=='
method for elements; now it returns false instead of an error if the
parents do not match. Added tests for both the '==' methods and the
binary methods (e.g. +, -, *) that require both arguments to have the
same parents.

SirToby25 committed:
Replaced length by ncols

Co-authored-by: Max Horn <max@quendi.de>

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Removed unnecessary usage of bitvectors

SirToby25 committed:
Compute j without log2

SirToby25 committed:
Renaming and single folder

SirToby25 committed:
Fixed basis_of_centroid and added tests

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
BoundsError thrown with i instead of 0

SirToby25 committed:
Removed third type variable for element types

joschmitt commented:
All the comments seem to be taken care of. Approve and merge?

thofma commented:
Why the ping? I was already happy, but @lgoettgens and @fingolfin had many comments and they should indicate whether they have been addressed.

lgoettgens commented:
My comments are all resolved, but I am not familiar enough with the mathematics to approve this.

joschmitt commented:
@fingolfin could you have a quick look again regarding your comments?

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordOrder.jl:
```diff
@@ -0,0 +1,1063 @@
+  parent::CliffordOrder{T, C}
+  coeffs::Vector{S} where S<:NumFieldElem
+  even_coeffs::Vector{S} where S<:NumFieldElem
+  odd_coeffs::Vector{S} where S<:NumFieldElem
```
This declaration means access to these three vectors will be type unstable.

lgoettgens replied:
This is fine to do (for reducing the number of method compilations) if you assert the concrete type *every* time you access these fields.

SirToby25 replied:
In this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/97234440e681122c3504c92e5df7344ab82fb6a0) I changed `Vector{S} where S<:NumFieldElem` to `Vector{<:NumFieldElem}`. Does this change anything?

lgoettgens replied:
No, this has no real difference. What you could do here is to add a type assertion to each place you access these vectors (ideally only in a getter)

SirToby25 replied:
As of this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/3e9dbfee8d0e6dfc2bd184bd9a9e04557505c52d), these attributes are of type `Any` but the concrete type is now asserted at access time

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+################################################################################
+
+### Algebra ###
+mutable struct CliffordAlgebra{T,S} <: Hecke.AbstractAssociativeAlgebra{T}
```
What are `T` and `S`? I would recommend to e.g. add a comment indicating what kind of types to expect here. People other than the original author who need to modify the code will probably grateful (based on experience that includes "the author in 1 year" ;-)

SirToby25 replied:
I added some clarifying comments in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/97234440e681122c3504c92e5df7344ab82fb6a0).

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+  dim::Int
+  basis_of_centroid::Any
+  disq::T
+  basis_of_center::Any
```
Lots of underspecified members (`base_ring`, `space`, `basis_of_centroid`, `basis_of_center`) meaning that code accessing them directly will be type unstable. You could compensate for that by using accessor function that have type assertions with the correct rings. It seems you already do so for `base_ring` and `space` but not for the others?

SirToby25 replied:
Addressed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/97234440e681122c3504c92e5df7344ab82fb6a0)

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+elem_type(::Type{CliffordAlgebra{T, S}}) where {T, S} =
+  CliffordAlgebraElem{T,S}
+
+  parent_type(::Type{CliffordAlgebraElem{T, S}}) where {T, S} = CliffordAlgebra{T, S}
```
```suggestion
parent_type(::Type{CliffordAlgebraElem{T, S}}) where {T, S} = CliffordAlgebra{T, S}
```

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+end
+
+elem_type(::Type{CliffordAlgebra{T, S}}) where {T, S} =
+  CliffordAlgebraElem{T,S}
```
```suggestion
elem_type(::Type{CliffordAlgebra{T, S}}) where {T, S} = CliffordAlgebraElem{T,S}
```

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+function gen(C::CliffordAlgebra, i::Int)
+  res = C()
+  if i <= 0
+    res[i] #Throws a BoundsError instead of a DomainError for consistency
```
```suggestion
    throw(BoundsError(res, i))
```

fingolfin commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,557 @@
+function setindex!(x::CliffordAlgebraElem, newentry::RingElement, i::Int64) 
+  coefficients(x)[i] = newentry
+end
+################################################################################
```
```suggestion

################################################################################
```

fingolfin reviewed (approved):
There are a bunch of type stability issue. But since this is code for experimental, I think we could merge it.

One other thing I'd like to see added to this eventually are "ring interface conformance tests". I.e. calls to `test_Ring_interface` or better yet, `test_Ring_interface_recursive`. This may require adding `test_elem` methods.

SirToby25 committed:
white space removed

Co-authored-by: Max Horn <max@quendi.de>

SirToby25 committed:
Line break removed

Co-authored-by: Max Horn <max@quendi.de>

SirToby25 committed:
Looks nicer now

Co-authored-by: Max Horn <max@quendi.de>

SirToby25 committed:
One more line

Co-authored-by: Max Horn <max@quendi.de>

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Clarified type variables, improved type stability

SirToby25 committed:
Fixed documention of base_ring

lgoettgens commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,563 @@
+two elements. The first one is the multiplicative identity of $C$. The square of
+the second basis element, if present, equals `quadratic_discriminant(C)`.
+"""
+function basis_of_centroid(C::CliffordAlgebra)::Vector{elem_type(C)}
```
```suggestion
function basis_of_centroid(C::CliffordAlgebra)
```
This lowers to a type conversion. To get a type assertion, instead put this to each line with a return, eg here `return C.basis_of_centroid::Vector{elem_type(C)}`. (at all places you just modified

SirToby25 replied:
I did not know that, thanks. It should work now, due to this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/3e9dbfee8d0e6dfc2bd184bd9a9e04557505c52d)

SirToby25 committed:
Type assertions now correctly placed at return

SirToby25 committed:
Merge branch 'oscar-system:master' into tb/CliffordAlgebras

SirToby25 committed:
Merge branch 'tb/CliffordAlgebras'

lgoettgens commented on experimental/CliffordAlgOrd/src/CliffordAlgebra.jl:
```diff
@@ -0,0 +1,563 @@
+"""
+function even_coefficients(x::CliffordAlgebraElem)
+  if isdefined(x, :even_coeffs)
+    return x.even_coeffs
```
This eg needs a type assertion as well (and 2 times below, and the other functions here)

SirToby25 replied:
Really? I thought the [type declarations of the data structure](https://github.com/oscar-system/Oscar.jl/pull/4370/files/f88323ea5a13e62d73c1ce5479faa7d82fc88d1a#diff-f830cdbcac3fbb51f8f7c3ae24d7af50985ce58653736e6ae50fc05943170069R45-R47) would be sufficient in this case.

lgoettgens replied:
Yes they are. I didn't notice that you changed that up there

lgoettgens commented on experimental/CliffordAlgOrd/src/CliffordOrder.jl:
```diff
@@ -0,0 +1,1058 @@
+
+Return the vector of coefficient ideals of the canonical pseudo-basis of $C$.
+"""
+coefficient_ideals(C::CliffordOrder) = C.coefficient_ideals::Vector{typeof(fractional_ideal(base_ring(C), one(base_ring(C))))}
```
I believe there is fractional_ideal_type to get this type at compile time (ie without typeof). Maybe you need to write Hecke.fractional_ideal_type (not sure, maybe @thofma knows)

thofma replied:
`Hecke.fractional_ideal_type(base_ring(C))` should do the trick.

lgoettgens replied:
Even better to use `Hecke.fractional_ideal_type(base_ring_type(C))` to make it compile time computable

SirToby25 replied:
Yes, this works

SirToby25 committed:
Fractional ideal type computable at compile time

SirToby25 committed:
Added various methods on elements

This commit adds many generic methods for elements of Clifford orders
that are not automatically available from AbstractAssociativeAlgebra.
Examples are: minimal_polynomial, characteristic_polynomial, inv, etc.

SirToby25 committed:
small fix - charpol, minpol

lgoettgens commented:
I think all comments are resovled now. And after all, this only adds experimental code, so it can just we revised later. Let's get this in. And thanks again to @SirToby25 for putting work into this!
